fix as_strided_scatter_backward

bdhirsh opened the pull request:
as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.

Fixes https://github.com/pytorch/pytorch/issues/88105

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88198
* #88230
* #88197
* #87647
* __->__ #87646
* #87255



cc @albanD

bdhirsh committed:
fix as_strided_scatter_backward

[ghstack-poisoned]

albanD reviewed (approved):
Great catch!
Also I guess this means we don't have opinfo (and thus autograd testing) for these ops... We should add them!

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.




[ghstack-poisoned]

bdhirsh commented:
we had a few skipped's OpInfo's for `as_strided_scatter` that I just cleaned up. There are still a few more, but... the OpInfo for `as_strided()` has some skips as well.

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.




cc albanD

[ghstack-poisoned]

albanD commented on test/functorch/test_aotdispatch.py:
```diff
@@ -952,7 +952,6 @@ def assert_compiler(gm: torch.fx.GraphModule, _):
     xfail('linalg.eig'),
     xfail('scatter_reduce', 'prod'),
 
-    # non-deterministic
```
Why remove this?

bdhirsh replied:
I thought the non-determinism was caused by the uninitialized memory 😛

mruberry replied:
Isn't the non-determinism that there may be multiple values mapping to the same index in the result, and which value appears there is nondeterministic? cc @kurtamohler

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.

Fixes https://github.com/pytorch/pytorch/issues/88105




cc albanD

[ghstack-poisoned]

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.

Fixes https://github.com/pytorch/pytorch/issues/88105




cc albanD

[ghstack-poisoned]

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.

Fixes https://github.com/pytorch/pytorch/issues/88105




cc albanD

[ghstack-poisoned]

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.

Fixes https://github.com/pytorch/pytorch/issues/88105




cc albanD

[ghstack-poisoned]

bdhirsh committed:
Update on "fix as_strided_scatter_backward"

as_strided_scatter's derivative formula was broken - instead of making a "mask" of 1's and 0's, it would effectively make a mask of 1's and uninitialized memory.

Fixes https://github.com/pytorch/pytorch/issues/88105




cc albanD

[ghstack-poisoned]

mruberry commented:
If this goes in after https://github.com/pytorch/pytorch/pull/85583/files then it should review the skips on the OpInfo

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR but I think this one or one of the PR in the stack break bionic-cuda11.7 on trunk https://hud.pytorch.org/pytorch/pytorch/commit/70782981f06a042796d4604df2ec1491f4f5b194' -c nosignal
